### PR TITLE
Support service account override

### LINKED ISSUE #448: Spark driver fails with K8s v1.6 out-of-box
Crashes with the following exception:

> Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://kubernetes.default.svc/api/v1/namespaces/default/pods/spark-dfsrw-1503358007989-driver. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. User "system:serviceaccount:default:default" cannot get pods in the namespace "default"..
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:332)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:269)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:241)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:234)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleGet(OperationSupport.java:230)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleGet(BaseOperation.java:745)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.get(BaseOperation.java:194)
	at org.apache.spark.scheduler.cluster.kubernetes.KubernetesClusterSchedulerBackend.liftedTree1$1(KubernetesClusterSchedulerBackend.scala:142)

It seems this is because the `default` service account does not have pod `edit` privilege any more under RBAC of K8s 1.6. Cluster admins are supposed to give only the barebone capability to the `default` account.

The recommendation is to add another service account, say `spark`, then add the necessary privileges:
```
$ kubectl create serviceaccount spark
$ kubectl create clusterrolebinding spark-edit --clusterrole edit --serviceaccount default:spark --namespace default
```

I did the above and I was hoping to use `--conf spark.kubernetes.authenticate.driver.serviceAccountName=spark`. But it seems the config key is no longer supported?

@foxish @mccheah

### PR DESCRIPTION
**Pull request by kimoonkim**
Fixes #448.

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/DriverKubernetesCredentialsStep.scala [hunk 1]
```diff
@@ -81,7 +82,17 @@ private[spark] class DriverKubernetesCredentialsStep(
...
         .build()
-    }.getOrElse(driverSpec.driverPod)
+    }.getOrElse(
+      driverServiceAccount.map {
+        account =>
+          new PodBuilder(driverSpec.driverPod)
+            .editOrNewSpec()
+            .withServiceAccount(account)
+            .withServiceAccountName(account)
+            .endSpec()
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/DriverKubernetesCredentialsStep.scala]: indent these in a couple spaces for each sub-item.  Even though the methods are all straight chained, there's a structure we've been trying to keep clear through indentation

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/DriverKubernetesCredentialsStep.scala [hunk 2]
```diff
@@ -81,7 +82,17 @@ private[spark] class DriverKubernetesCredentialsStep(
             .endVolume()
           .endSpec()
         .build()
-    }.getOrElse(driverSpec.driverPod)
+    }.getOrElse(
+      driverServiceAccount.map {
+        account =>
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/DriverKubernetesCredentialsStep.scala]: move account onto previous line to match style in rest of this file

**Review by ash211 — COMMENTED** [review]
Seems generally right, though I think I remember @foxish having an opinion about service accounts previously.

@mccheah is this the right step to make this change in?

**Comment by erikerlandson** [comment]
This logic implements a policy that kube secret credentials override service acct, and also prevent service acct from being applied.  Is that the right policy?

**Comment by mccheah** [comment]
I think this is the right policy.

**Comment by ash211** [comment]
I thought we had validation that you can't apply both secret creds and a service account -- you have to pick one or the other

**Comment by mccheah** [comment]
I think either is fine really, but we should have preference for the non-service account based authentication.

**Comment by erikerlandson** [comment]
I was imagining what amounts to supplementing permissions from one with the other, but that might be a corner case.

**Comment by erikerlandson** [comment]
It sounds like the current logic is restoring the original logic that was clobbered - LGTM

**Comment by kimoonkim** [comment]
It seems everybody is fine with this, although it isn't approved yet. Maybe we can approve and merge this soon. I also wonder if we need to have a dot release in case more people start to use Kubernetes 1.6.

**Comment by erikerlandson** [comment]
@foxish, you're assigned, do you want to merge?

**Comment by foxish** [comment]
LGTM

**Comment by kimoonkim** [comment]
Thanks everyone for the reviews!

**Pull request merged**